### PR TITLE
Feature/ellen/add corona stats il

### PR DESCRIPTION
**Pull request by ellenc345**

### CODE REVIEW

**Review by subscorp — COMMENTED** [review]
Looks really good :)

File: tests/test_corona_stats.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    assert False
+
+
+@patch('requests.get')
+def test_get_vacinated_data(mock_get):
+    test_data = json.dumps(FAKE_DATA)
+    mock_get.return_value.ok = True
+    mock_get.return_value.text = test_data
+    data = corona_stats.get_vacinated_data()
+    assert data is not None
```
Comment (subscorp) [inline tests/test_corona_stats.py]: assert data is sufficient.

File: tests/test_corona_stats.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+    assert is_empty(session) is False
+
+
+@patch('requests.get')
+def test_get_corona_stats(mock_get, session):
+    test_data = json.dumps(FAKE_DATA)
+    mock_get.return_value.ok = True
+    mock_get.return_value.text = test_data
+    assert is_empty(session) is True
```
Comment (subscorp) [inline tests/test_corona_stats.py]: If is_empty(session) returns bool, you can write "assert is empty(session)". No need for the "is True" part.

File: tests/test_corona_stats.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+    data = corona_stats.get_corona_stats(session)
+    assert data is not None
+
+    assert is_empty(session) is False
```
Comment (subscorp) [inline tests/test_corona_stats.py]: Same as line 87. Prefer "assert not is_empty(session)".

**Review by subscorp — CHANGES_REQUESTED** [review]

File: app/database/models.py [hunk 4]
```diff
@@ -290,6 +291,20 @@ class WikipediaEvents(Base):
...
+class CoronaStats(Base):
+    __tablename__ = "corona_stats"
+
+    id = Column(Integer, primary_key=True, index=True)
+    date_ = Column(DateTime, nullable=False)
+    date_inserted = Column(DateTime, default=datetime.utcnow)
+    vaccinated = Column(Integer, nullable=False)
+    vaccinated_cum = Column(Integer, nullable=False)
+    vaccinated_population_perc = Column(Integer, nullable=False)
+    vaccinated_seconde_dose = Column(Integer, nullable=False)
```
Comment (yammesicka) [inline app/database/models.py]: second

File: app/database/models.py [hunk 5]
```diff
@@ -290,6 +291,20 @@ class WikipediaEvents(Base):
...
+    vaccinated_seconde_dose_cum = Column(Integer, nullable=False)
+    vaccinated_seconde_dose_population_perc = Column(Float, nullable=False)
```
Comment (yammesicka) [inline app/database/models.py]: Remove `population` to keep it short

File: app/database/models.py [hunk 6]
```diff
@@ -290,6 +291,20 @@ class WikipediaEvents(Base):
...
 
+class CoronaStats(Base):
+    __tablename__ = "corona_stats"
+
+    id = Column(Integer, primary_key=True, index=True)
+    date_ = Column(DateTime, nullable=False)
+    date_inserted = Column(DateTime, default=datetime.utcnow)
+    vaccinated = Column(Integer, nullable=False)
+    vaccinated_cum = Column(Integer, nullable=False)
+    vaccinated_population_perc = Column(Integer, nullable=False)
```
Comment (yammesicka) [inline app/database/models.py:453]: Remove `population` to keep it short

File: app/internal/corona_stats.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+import requests
+from fastapi import Depends
+from loguru import logger
+from sqlalchemy import desc, func
+from sqlalchemy.exc import SQLAlchemyError
+from sqlalchemy.orm import Session
+from sqlalchemy.orm.exc import NoResultFound
+from sqlalchemy.sql.sqltypes import JSON
+
```
Comment (yammesicka) [inline app/internal/corona_stats.py]: abc

File: app/internal/corona_stats.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+        ),
+        vaccinated_seconde_dose=data.get(
+            "vaccinated_seconde_dose"),
+        vaccinated_seconde_dose_cum=data.get(
+            "vaccinated_seconde_dose_cum"
+        ),
+        vaccinated_seconde_dose_population_perc=data.get(
+            "vaccinated_seconde_dose_population_perc"
+        )
+    )
```
Comment (yammesicka) [inline app/internal/corona_stats.py]: Use git precommit hooks to reformat

File: app/internal/corona_stats.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+    return {
+        "vaccinated_seconde_dose_population_perc":
+        stats_object.vaccinated_seconde_dose_population_perc,
+        "vaccinated_seconde_dose_cum": stats_object.vaccinated_seconde_dose_cum
+    }
+
+
+def save_corona_stats(data: JSON, db: Session = Depends(get_db)) -> None:
+    db.add(
+        create_stats_object(data)
```
Comment (yammesicka) [inline app/internal/corona_stats.py]: Can be in one line

File: app/internal/corona_stats.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+        if latest_saved.date_ < latest_date:
+            save_corona_stats(data, db)
+            return data
+        else:
+            return serialize_stats(latest_saved)
+
+
+def get_vacinated_data() -> JSON:
+    # the api updates during the day, so we want to stay updated.
+    res = requests.get(CORONA_API_URL)
```
Comment (yammesicka) [inline app/internal/corona_stats.py]: Use httpx for async requests

File: app/internal/corona_stats.py [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+    try:
+        db_data = get_vacinated_data_from_db(db)
+        data = serialize_stats(db_data)
+
+    except NoResultFound:
+        data = get_vacinated_data()
+        insert_to_db_if_needed(data, db)
+
+    except (SQLAlchemyError, AttributeError) as e:
+        logger.error(f'corona stats failed with error: {e}')
```
Comment (yammesicka) [inline app/internal/corona_stats.py]: Prefer .exception

File: app/internal/corona_stats.py [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+    return (db.query(CoronaStats).
+            filter(
+            func.date(CoronaStats.date_) == date.today()).
+            one())
+
+
+def get_corona_stats(db: Session = Depends(get_db)) -> Dict[str, Any]:
+    try:
+        db_data = get_vacinated_data_from_db(db)
+        data = serialize_stats(db_data)
```
Comment (yammesicka) [inline app/internal/corona_stats.py]: Please use a better variable names (avoid using the word `data`)

**Review by yammesicka — COMMENTED** [review]
Great start! I've added few insights :)

File: app/database/models.py (hunk 6 above)
Comment (ellenc345) [reply app/database/models.py:453]: I wanted to keep the original data stracture , is it that bad?

File: tests/test_corona_stats.py [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+           filter().
+           count()
+           )
+    return res == 0
+
+
+def test_get_vacinated_data_from_db(session):
+    try:
+        corona_stats.get_vacinated_data_from_db(session)
+    except NoResultFound:
```
Comment (ivarshav) [inline tests/test_corona_stats.py]: You should test the "good" and "bad" cases separately

Comment (ellenc345) [reply tests/test_corona_stats.py]: what does it means? it testing only one case here, if there is data the test wil fail

Comment (ivarshav) [reply tests/test_corona_stats.py]: You probably want something like [this](https://docs.pytest.org/en/reorganize-docs/new-docs/user/pytest_raises.html)

**Comment by ellenc345** [comment]
Waiting for review :)

File: tests/test_corona_stats.py [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+
+import pytest
+from sqlalchemy.orm.exc import NoResultFound
+
+from app.database.models import CoronaStats
+from app.internal import corona_stats
+
+FAKE_DATA = [
+    {
+        "Day_Date": "2020-12-19T00:00:00.000Z",
```
Comment (Gonzom) [inline tests/test_corona_stats.py:11]: lowercase

File: app/database/models.py [hunk 15]
```diff
@@ -372,6 +373,20 @@ class WikipediaEvents(Base):
...
 
 
+class CoronaStats(Base):
+    __tablename__ = "corona_stats"
+
+    id = Column(Integer, primary_key=True, index=True)
+    date_ = Column(DateTime, nullable=False)
+    date_inserted = Column(DateTime, default=datetime.utcnow)
+    vaccinated = Column(Integer, nullable=False)
+    vaccinated_cum = Column(Integer, nullable=False)
```
Comment (Gonzom) [inline app/database/models.py]: What is "cum" short for? It's not very clear.

File: app/internal/corona_stats.py [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+        vaccinated_second_dose_cum=corona_stats_data.get(
+            "vaccinated_seconde_dose_cum",
+        ),
+        vaccinated_second_dose_perc=corona_stats_data.get(
+            "vaccinated_seconde_dose_population_perc",
+        ),
+    )
+
+
+def serialize_stats(stats_object: CoronaStats) -> Dict:
```
Comment (Gonzom) [inline app/internal/corona_stats.py]: Dict of what?

File: app/internal/corona_stats.py [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+            stats_dict.get("vaccinated_seconde_dose_population_perc")
+        ),
+        "vaccinated_second_dose_cum": (
+            stats_dict.get("vaccinated_seconde_dose_cum")
+        ),
+    }
+
+
+def save_corona_stats(
+    corona_stats_data: JSON,
```
Comment (Gonzom) [inline app/internal/corona_stats.py]: is the type(corona_stats_data) result of this a JSON object?

File: app/internal/corona_stats.py [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+        "vaccinated_second_dose_perc": (
+            stats_object.vaccinated_second_dose_perc
+        ),
+        "vaccinated_second_dose_cum": (
+            stats_object.vaccinated_second_dose_cum
+        ),
+    }
+
+
+def serialize_dict_stats(stats_dict: Dict) -> Dict:
```
Comment (Gonzom) [inline app/internal/corona_stats.py]: Dict of what?

File: app/internal/corona_stats.py [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+
+async def get_vacinated_data() -> JSON:
+    async with httpx.AsyncClient() as client:
+        res = await client.get(CORONA_API_URL, headers=headers)
+        return json.loads(res.text)[-1]
+
+
+def get_vacinated_data_from_db(db: Session = Depends(get_db)) -> CoronaStats:
+    # I'm pulling once a day, it won't be the most updated data
+    # but we dont want to be blocked for to many requests
```
Comment (Gonzom) [inline app/internal/corona_stats.py]: Don't write comments in the first person.

File: app/static/images/icons/israel.svg [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+<g>
+</g>
+<g>
+</g>
+<g>
+</g>
+<g>
+</g>
+<g>
+</g>
```
Comment (Gonzom) [inline app/static/images/icons/israel.svg:46]: is this valid and needed?

File: app/templates/corona_stats.html [hunk 21]
```diff
@@ -0,0 +1,16 @@
...
+{% if corona_stats_data.get("error") is none %}
+<div class="card card-profile">
+  <div class="card-body">
+    <div class="container">
+      <p class="card-title">
+        <img class="pb-1" src="{{ url_for('static', path='/images/icons/israel.svg')}}" width="20em" height="20em">
+        COVID 19 Vaccinated</p>
+      <p>
+        {{corona_stats_data["vaccinated_second_dose_perc"]}}%
+        ({{corona_stats_data["vaccinated_second_dose_cum"]}})
```
Comment (Gonzom) [inline app/templates/corona_stats.html]: There should be a space between "{{" and the variable.

File: app/templates/corona_stats.html [hunk 22]
```diff
@@ -0,0 +1,16 @@
+
+{% if corona_stats_data.get("error") is none %}
+<div class="card card-profile">
+  <div class="card-body">
+    <div class="container">
+      <p class="card-title">
+        <img class="pb-1" src="{{ url_for('static', path='/images/icons/israel.svg')}}" width="20em" height="20em">
```
Comment (Gonzom) [inline app/templates/corona_stats.html:7]: Space before "}}".

File: app/templates/corona_stats.html [hunk 23]
```diff
@@ -0,0 +1,16 @@
...
+        COVID 19 Vaccinated</p>
+      <p>
+        {{corona_stats_data["vaccinated_second_dose_perc"]}}%
+        ({{corona_stats_data["vaccinated_second_dose_cum"]}})
+      </p>
+    </div>
+  </div>
+</div>
+{% endif%}
```
Comment (Gonzom) [inline app/templates/corona_stats.html]: space

File: app/static/images/icons/israel.svg (hunk 20 above)
Comment (ellenc345) [reply app/static/images/icons/israel.svg:46]: This is israel icon

File: tests/test_corona_stats.py (hunk 14 above)
Comment (ellenc345) [reply tests/test_corona_stats.py:11]: This is constant, why lowercase?

File: app/internal/corona_stats.py (hunk 17 above)
Comment (ellenc345) [reply app/internal/corona_stats.py]: i dont understant the comment and what i should fix in the code, can you please explain ?

File: app/static/images/icons/israel.svg (hunk 20 above)
Comment (Gonzom) [reply app/static/images/icons/israel.svg:46]: I marked the opening and closing of the <g> tags. Why are these empty tags?

File: app/internal/corona_stats.py (hunk 17 above)
Comment (Gonzom) [reply app/internal/corona_stats.py]: In you code do print(type(corona_stats_data)). What type does it show there? I was just wondering if the type is actually JSON and not a Dict.

File: tests/test_corona_stats.py (hunk 14 above)
Comment (Gonzom) [reply tests/test_corona_stats.py:11]: Sorry if I wasn't clear, I wasn't talking about the constant but about "Day_Date" which should be "day_date".

File: app/internal/corona_stats.py (hunk 17 above)
Comment (ellenc345) [reply app/internal/corona_stats.py]: its formmating and saving the json from api to the database

Comment (Gonzom) [reply app/internal/corona_stats.py]: Not what I asked...

File: app/static/images/icons/israel.svg (hunk 20 above)
Comment (ellenc345) [reply app/static/images/icons/israel.svg:46]: this is the way the svg was created, im not an svg creator, just using it.

File: app/internal/corona_stats.py (hunk 17 above)
Comment (ellenc345) [reply app/internal/corona_stats.py]: its JSON loaded from the http response

File: app/internal/corona_stats.py [hunk 24]
```diff
@@ -0,0 +1,155 @@
...
+    "Gecko/20100101 Firefox/77.0",
+    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
+    "AppleWebKit/537.36 (KHTML, like Gecko)"
+    " Chrome/83.0.4103.97 Safari/537.36",
+]
+
+headers = {"User-Agent": random.choice(user_agent_list)}
+
+
+def create_stats_object(corona_stats_data: JSON) -> CoronaStats:
```
Comment (yammesicka) [inline app/internal/corona_stats.py]: That's not the correct type for `corona_stats_data`

File: app/internal/corona_stats.py [hunk 25]
```diff
@@ -0,0 +1,155 @@
...
+    db: Session = Depends(get_db),
+) -> None:
+    db.add(create_stats_object(corona_stats_data))
+    db.commit()
+
+
+def insert_to_db_if_needed(
+    corona_stats_data: JSON,
+    db: Session = Depends(get_db),
+):
```
Comment (yammesicka) [inline app/internal/corona_stats.py]: Add return type

File: app/internal/corona_stats.py [hunk 26]
```diff
@@ -0,0 +1,155 @@
...
+    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
+    "AppleWebKit/537.36 (KHTML, like Gecko)"
+    " Chrome/83.0.4103.97 Safari/537.36",
+]
+
+headers = {"User-Agent": random.choice(user_agent_list)}
+
+
+def create_stats_object(corona_stats_data: JSON) -> CoronaStats:
+    """ JSON -> DB Object """
```
Comment (yammesicka) [inline app/internal/corona_stats.py]: Remove spaces

File: app/internal/corona_stats.py [hunk 27]
```diff
@@ -0,0 +1,155 @@
...
+    if latest_saved is not None:
+        # on more recent data arrival, we update the database
+        if latest_saved.date_ < latest_date:
+            save_corona_stats(corona_stats_data, db)
+            return corona_stats_data
+        else:
+            return serialize_stats(latest_saved)
+
+
+async def get_vacinated_data() -> JSON:
```
Comment (yammesicka) [inline app/internal/corona_stats.py]: Not the correct return type

File: app/routers/profile.py [hunk 28]
```diff
@@ -27,20 +31,21 @@
...
 
 
 @router.get("/")
 async def profile(
-        request: Request,
-        session=Depends(get_db),
-        new_user=Depends(get_placeholder_user)):
+    request: Request,
+    session=Depends(get_db),
+    new_user=Depends(get_placeholder_user),
```
Comment (yammesicka) [inline app/routers/profile.py]: Use user from Kobi's login system

File: app/internal/corona_stats.py [hunk 29]
```diff
@@ -0,0 +1,155 @@
...
+    "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:77.0)"
+    " Gecko/20100101 Firefox/77.0",
+    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_5) "
+    "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.97 Safari/537.36",
+    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:77.0) "
+    "Gecko/20100101 Firefox/77.0",
+    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
+    "AppleWebKit/537.36 (KHTML, like Gecko)"
+    " Chrome/83.0.4103.97 Safari/537.36",
+]
```
Comment (yammesicka) [inline app/internal/corona_stats.py]: Why do we need this? From quick trial the API plays nice with Python

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: app/internal/corona_stats.py (hunk 29 above)
Comment (ellenc345) [reply app/internal/corona_stats.py]: At some point I started to get 403 from the code while it worked fine from the browser, I get it to work again only with different user agents

File: app/routers/profile.py (hunk 28 above)
Comment (ellenc345) [reply app/routers/profile.py]: I'm afraid this code is reletedto someone elses feature, im not using the user parameter

File: app/internal/corona_stats.py (hunk 29 above)
Comment (yammesicka) [reply app/internal/corona_stats.py]: You use only a single user agent from the list. Maybe move the `headers` to be inside of the functions

**Review by yammesicka — CHANGES_REQUESTED** [review]
Fix small problem and DM me to merge this PR

File: app/internal/corona_stats.py (hunk 29 above)
Comment (ellenc345) [reply app/internal/corona_stats.py]: oh! you are right! i misse that.

**Pull request merged**